### PR TITLE
[TIMOB-25313] iOS: Add parity with Android

### PR DESCRIPTION
**Pull request by hansemannn**
Add parity with Android by using the new iOS 11 `CoreNFC` system-framework. Basic usage is done, docs still need to be updated and some properties need to be tested against type-parity.

**JIRA**: https://jira.appcelerator.org/browse/TIMOB-25313

### CODE REVIEW

File: apidoc/NdefRecord.yml [hunk 1]
```diff
@@ -10,12 +12,17 @@ description: |
...
+    to read specific NDEF-tags - writing tags is not supported by iOS so far.
 
     See also:
     [NdefRecord](http://developer.android.com/reference/android/nfc/NdefRecord.html)
 extends: Titanium.Proxy
-since: "1.0.0"
-platforms: [android]
+since: {android: "1.0.0", iphone: "6.3.0", ipad: "6.3.0"}
+osver: {ios: {min: "11.0"}}
+platforms: [android, iphone, ipad]
```
Comment (vijaysingh-axway) [inline apidoc/NdefRecord.yml]: As I see at other places in document, only 'iphone' is supported. I think we should remove' ipad' from above as well.

File: ios/Classes/TiNfcNfcAdapterProxy.m [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+#pragma mark NFCNDEFReaderSessionDelegate
+
+- (void)readerSession:(NFCNDEFReaderSession *)session didDetectNDEFs:(NSArray<NFCNDEFMessage *> *)messages
+{
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity:[messages count]];
+
+  for (NFCNDEFMessage *message in messages) {
+    [result addObject:[[TiNfcNdefMessageProxy alloc] _initWithPageContext:[self pageContext]
```
Comment (vijaysingh-axway) [inline ios/Classes/TiNfcNfcAdapterProxy.m]: What is use of result? It is not getting used anywhere in this function.

File: apidoc/NdefRecord.yml (hunk 1 above)
Comment (hansemannn) [reply apidoc/NdefRecord.yml]: Good catch!

File: ios/Classes/TiNfcNfcAdapterProxy.m (hunk 2 above)
Comment (hansemannn) [reply ios/Classes/TiNfcNfcAdapterProxy.m]: It is returned in the same method and `- (NSArray<TiNfcNdefRecordProxy *> *)records` returns to the proxy as `records`.

Comment (vijaysingh-axway) [reply ios/Classes/TiNfcNfcAdapterProxy.m]: I think you are talking about TiNfcNdefMessageProxy class. But above method is delegate method of TiNfcNfcAdapterProxy class. You are passing -
[_ndefDiscoveredCallback call:@[ @{@"messages" : [TiNfcUtilities arrayFromNDEFMessages:messages]} ]            thisObject:self];
Corresponding to "messages"  key I think you have to pass result. If I understand right, I think there is no use of TiNfcUtilities class. Can you please once review  delegate method -
+- (void)readerSession:(NFCNDEFReaderSession *)session didDetectNDEFs:(NSArray<NFCNDEFMessage *> *)messages

of  TiNfcNfcAdapterProxy class.  Thanks!

Comment (hansemannn) [reply ios/Classes/TiNfcNfcAdapterProxy.m]: Checking it, the loop is indeed not used, it was moved to the utilities. But the `TiNfcUtilities` are still used to bridge the messages to proxies. I updated the PR to remove the unused code!

Comment (vijaysingh-axway) [reply ios/Classes/TiNfcNfcAdapterProxy.m]: If you process the messages received in  callback using TiNfcUtilities then TiNfcNdefMessageProxy  and TiNfcNdefRecordProxy class is of no use.

Comment (hansemannn) [reply ios/Classes/TiNfcNfcAdapterProxy.m]: You are right! Fixed that.

**Comment by vijaysingh-axway** [comment]
Note: I couldn't test the message reading of tag at my end as it requires nfc tag.

**Comment by drunnells** [comment]
Hopefully this is an OK place to report this - I am unable to read the payload of an NFCs with an iPhone 7 with ios11 using this module. The module and example app.js build fine with xcode 9 and v6.2.2 of the Titanium SDK and my provisioning profile and entitlements seem to be OK, but when I build and install the example app.js and scan an NFC card that I've written an NDEF record to I receive warnings and an empty payload. I've written a URL to the NDEF record ("somewhere.com/something") and can read this with the android ti.nfc module in another application. But for the new ios version I get the below warnings and nothing in the payload:

```
[INFO] :   NFC Test/1.0 (6.2.2.42c7196)
[WARN] :   Creating [object TiNfcNdefMessage] in a different context than the calling function.
[WARN] :   {
[WARN] :       messages =     (
[WARN] :           "<null>"
[WARN] :       );
[WARN] :   }
```

Prior to the 00016e6 commit I was getting similar results, but the warning contained the URL from the NDEF record:
```
[INFO] :   NFC Test/1.0 (6.2.2.42c7196)
[WARN] :   Creating somewhere.com/something in a different context than the calling function.
[WARN] :   {
[WARN] :       messages =     (
[WARN] :                   {
[WARN] :               records =             (
[WARN] :                                   {
[WARN] :                       id = "";
[WARN] :                       payload = "<null>";
[WARN] :                       tnf = 1;
[WARN] :                       type = U;
[WARN] :                   }
[WARN] :               );
[WARN] :           }
[WARN] :       );
[WARN] :   }
```

If it helps, using an NFC Taginfo app on Android I see the card as "Type A (ISO/IEC 14443 Type A)" with Mifare Classic as NDEF Tag:

WELL-KNOWN: urn:nfc:wkt:U (URI)
Identifier: 0x01 ("http://www.")
"somewhere.com/something"

**Comment by hansemannn** [comment]
@drunnells Thanks, reverted that commit for now. I feel like that tag needs a special data-decoding, but I do not have an URI-Tag right now. I'll try to reprogram my existing one.

**Comment by drunnells** [comment]
@hansemannn I tried a few more tags, but get the same results, even when the message is just text. I'd be happy to test other tags, configurations, or NDEF messages if you need. I thought that the problem was more related to the "Creating in a different context" warning? Here are the other tags that I have tried:

ISO 14443-A
NXP MIFARE Ultralight (Ultralight C)
Data format: NFC Forum Type 2

ISO 14443-3A
NXP MIFARE Classic 1k
Data format: NXP Mifare Classic

ISO 14443-3A
NXP MIFARE Ultralight (Ultralight)
Data format: NFC FOrum Type 2

**Comment by hansemannn** [comment]
Hey there! I drafted a new beta that uses [this decoding](https://stackoverflow.com/a/44446471/5537752) for the payload and type. Can you try it out?

[ti.nfc-iphone-2.0.0.zip](https://github.com/appcelerator-modules/ti.nfc/files/1371806/ti.nfc-iphone-2.0.0.zip)

**Comment by drunnells** [comment]
@hansemannn

I gave it a try, but I'm afraid that I did not notice any change in the results with the example app.js - messages seems to be null still for both URL and Text NDEF records. Hopefully I'm not missing anything obvious.. I'd think that if I had my entitlements or profile setup incorrectly I wouldn't even be getting this far.

```
Please manually launch the application
[INFO] :   NFC Test/1.0 (6.2.2.42c7196)
[WARN] :   Creating [object TiNfcNdefMessage] in a different context than the calling function.
[WARN] :   {
[WARN] :       messages =     (
[WARN] :           "<null>"
[WARN] :       );
[WARN] :   }
[ERROR] :  Script Error {
[ERROR] :      column = 186;
[ERROR] :      line = 1;
[ERROR] :      message = "null is not an object (evaluating 'c.records')";
[ERROR] :      sourceURL = "file:///var/containers/Bundle/Application/6821C8CF-996C-4395-B953-153CE3D98FCA/NFC%20Test.app/app.js";
[ERROR] :      stack = "handleDiscovery@file:///var/containers/Bundle/Application/6821C8CF-996C-4395-B953-153CE3D98FCA/NFC%20Test.app/app.js:1:186";
[ERROR] :  }
```

I also tried running the project with Xcode to try and provide you with additional debugging info, not sure how useful it will be to you:

```
2017-10-10 22:52:13.460225-0400 NFC Test[1099:2768312] [DYMTLInitPlatform] platform initialization successful
2017-10-10 22:52:13.522709-0400 NFC Test[1099:2768108] [DEBUG] Reading stylesheet from: /var/containers/Bundle/Application/D626FC92-7C9E-4CA1-A582-4032A2958ADE/NFC Test.app/stylesheet.plist
2017-10-10 22:52:13.778021-0400 NFC Test[1099:2768108] [INFO] NFC Test/1.0 (6.2.2.42c7196)
2017-10-10 22:52:13.868418-0400 NFC Test[1099:2768334] [DEBUG] Loading: /var/containers/Bundle/Application/D626FC92-7C9E-4CA1-A582-4032A2958ADE/NFC Test.app/app.js, Resource: app_js
2017-10-10 22:52:13.868844-0400 NFC Test[1099:2768108] refreshPreferences: HangTracerEnabled: 0
2017-10-10 22:52:13.868907-0400 NFC Test[1099:2768108] refreshPreferences: HangTracerDuration: 500
2017-10-10 22:52:13.868946-0400 NFC Test[1099:2768108] refreshPreferences: ActivationLoggingEnabled: 0 ActivationLoggingTaskedOffByDA:0
2017-10-10 22:52:13.889291-0400 NFC Test[1099:2768334] [DEBUG] [object TiNfcModule] loaded
=================================================================
Main Thread Checker: UI API called on a background thread: -[UIView superview]
PID: 1099, TID: 2768334, Thread name: KrollContext<kroll$1>, Queue name: com.apple.root.default-qos.overcommit, QoS: 21
Backtrace:
4   NFC Test                            0x0000000100d7b878 NFC Test + 981112
5   NFC Test                            0x0000000100d7ba18 NFC Test + 981528
6   CoreFoundation                      0x00000001859c76a0 <redacted> + 144
7   CoreFoundation                      0x00000001858a6820 <redacted> + 292
8   NFC Test                            0x0000000100ca86e8 NFC Test + 116456
9   NFC Test                            0x0000000100ca6cb0 NFC Test + 109744
10  NFC Test                            0x0000000100e75244 NFC Test + 2003524
11  NFC Test                            0x0000000100f14da4 NFC Test + 2657700
12  NFC Test                            0x0000000100f1acd8 NFC Test + 2682072
2017-10-10 22:52:13.908249-0400 NFC Test[1099:2768334] [reports] Main Thread Checker: UI API called on a background thread: -[UIView superview]
PID: 1099, TID: 2768334, Thread name: KrollContext<kroll$1>, Queue name: com.apple.root.default-qos.overcommit, QoS: 21
Backtrace:
4   NFC Test                            0x0000000100d7b878 NFC Test + 981112
5   NFC Test                            0x0000000100d7ba18 NFC Test + 981528
6   CoreFoundation                      0x00000001859c76a0 <redacted> + 144
7   CoreFoundation                      0x00000001858a6820 <redacted> + 292
8   NFC Test                            0x0000000100ca86e8 NFC Test + 116456
9   NFC Test                            0x0000000100ca6cb0 NFC Test + 109744
10  NFC Test                            0x0000000100e75244 NFC Test + 2003524
11  NFC Test                            0x0000000100f14da4 NFC Test + 2657700
12  NFC Test                            0x0000000100f1acd8 NFC Test + 2682072
2017-10-10 22:52:16.914786-0400 NFC Test[1099:2768108] [DEBUG] Loading: /var/containers/Bundle/Application/D626FC92-7C9E-4CA1-A582-4032A2958ADE/NFC Test.app/_app_props_.json, Resource: _app_props__json
2017-10-10 22:52:16.917005-0400 NFC Test[1099:2768334] [DEBUG] Application booted in 3296.485007 ms
2017-10-10 22:52:28.222275-0400 NFC Test[1099:2768322] [WARN] Creating [object TiNfcNdefMessage] in a different context than the calling function.
2017-10-10 22:52:28.229231-0400 NFC Test[1099:2768322] [WARN] {
    messages =     (
        "<null>"
    );
}
2017-10-10 22:52:28.230975-0400 NFC Test[1099:2768322] [ERROR] Script Error {
    column = 186;
    line = 1;
    message = "null is not an object (evaluating 'c.records')";
    sourceURL = "file:///var/containers/Bundle/Application/D626FC92-7C9E-4CA1-A582-4032A2958ADE/NFC%20Test.app/app.js";
    stack = "handleDiscovery@file:///var/containers/Bundle/Application/D626FC92-7C9E-4CA1-A582-4032A2958ADE/NFC%20Test.app/app.js:1:186";
}
```

**Comment by hansemannn** [comment]
@drunnells Out of interest, does it make a change to use `run-on-main-thread` in your tiapp.xml?

**Comment by drunnells** [comment]
@hansemannn No change after adding:

`<property name="run-on-main-thread" type="bool">true</property>`

to the ios section. Same thing for false.

```
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
    <id>...</id>
    <name>...</name>
    <version>1.0</version>
    <publisher>...</publisher>
    <url>...</url>
    <description>undefined</description>
    <copyright>...</copyright>
    <icon>appicon.png</icon>
    <fullscreen>true</fullscreen>
    <navbar-hidden>true</navbar-hidden>
    <analytics>false</analytics>
    <guid>...</guid>
    <property name="ti.ui.defaultunit" type="string">dp</property>
    <ios>
        <property name="run-on-main-thread" type="bool">true</property>
        <enable-launch-screen-storyboard>false</enable-launch-screen-storyboard>
        <use-app-thinning>true</use-app-thinning>
		<entitlements>
			<dict>
				<key>com.apple.developer.nfc.readersession.formats</key>
				<array>
					<string>NDEF</string>
				</array>
			</dict>
		</entitlements>
        <plist>
            <dict>
                <key>UISupportedInterfaceOrientations~iphone</key>
                <array>
                    <string>UIInterfaceOrientationPortrait</string>
                </array>
                <key>UISupportedInterfaceOrientations~ipad</key>
                <array>
                    <string>UIInterfaceOrientationPortrait</string>
                    <string>UIInterfaceOrientationPortraitUpsideDown</string>
                    <string>UIInterfaceOrientationLandscapeLeft</string>
                    <string>UIInterfaceOrientationLandscapeRight</string>
                </array>
                <key>UIRequiresPersistentWiFi</key>
                <false/>
                <key>UIPrerenderedIcon</key>
                <false/>
                <key>UIStatusBarHidden</key>
                <false/>
                <key>UIStatusBarStyle</key>
                <string>UIStatusBarStyleDefault</string>
                <key>NFCReaderUsageDescription</key>
                <string>NFC Tag!</string>
                <key>NSContactsUsageDescription</key>
                <string>Can we use to your contacts?</string>
                <key>NSCameraUsageDescription</key>
                <string>Can we use your camera?</string>
                <key>NSCalendarsUsageDescription</key>
                <string>Can we use your calendar?</string>
                <key>NSPhotoLibraryUsageDescription</key>
                <string>Can we save to your library?</string>
                <key>NSMicrophoneUsageDescription</key>
                <string>Can we use your microphone?</string>
            </dict>
        </plist>
    </ios>
    <android xmlns:android="http://schemas.android.com/apk/res/android"/>
    <mobileweb>
        <precache/>
        <splash>
            <enabled>true</enabled>
            <inline-css-images>true</inline-css-images>
        </splash>
        <theme>default</theme>
    </mobileweb>
    <modules>
        <module platform="iphone">ti.nfc</module>
    </modules>
    <deployment-targets>
        <target device="android">false</target>
        <target device="ipad">false</target>
        <target device="iphone">true</target>
        <target device="mobileweb">false</target>
    </deployment-targets>
    <sdk-version>6.2.2.GA</sdk-version>
    <property name="appc-app-id"/>
</ti:app>
```

**Comment by hansemannn** [comment]
Ok, that won't be it. But you should use `  <property name="run-on-main-thread" type="bool">true</property>` instead of the <ios> one.

**Comment by eddyinet** [comment]
Any luck decoding the NDEF message on ios? I'm still getting:

`Creating [object TiNfcNdefMessage] in a different context than the calling function.
[WARN] :   {
[WARN] :       messages =     (
[WARN] :           "<null>"
[WARN] :       );
[WARN] :   }`

**Pull request merged**